src/content/en/fundamentals/getting-started/primers/media-source-extensions.md

```diff
@@ -0,0 +1,323 @@
+function sourceOpen(e) {  
+  var mime = 'video/webm; codecs="opus, vp9"';  
+  var mediaSource = this;  
+  var sourceBuffer = mediaSource.addSourceBuffer(mime);  
+  var video = 'droid.webm'; 
+  <strong>fetch(video)
+    .then(function(response) {
+      // Process the response object.
+    });</strong>
+  }
+}</pre>
+
+A production quality player would have the same file in multiple versions to
+support different browsers. It could use separate files for audio and video to
+allow audio to be selected based on language settings.
+
+Real world code would also have multiple copies of media files at different
+resolutions so that it could adapt to different device capabilities and network
+conditions. Such an application is able to load and play videos in chunks either
+using range requests or segments. This allows for adaption to network conditions
+*while media are playing*. You may have heard the terms DASH or HLS, which are
+two methods of accomplishing this. A full discussion of this topic is beyond the
+scope of this introduction.
+
+## Process the response object
+
+The code looks almost done, but the media doesn't play. We need to get media
+data from the `Response` object to the `SourceBuffer`.
+
+The typical way to pass data from the response object to the `MediaSource`
+instance is to get an `ArrayBuffer` from the response object and pass it to the
+`SourceBuffer`. Start by calling `response.arrayBuffer()`, which returns a
+promise to the buffer. In my code, I've passed this promise to a second `then()`
+clause where I append it to the `SourceBuffer`.
+
+<pre class="prettyprint">
+function sourceOpen(e) {
+  var mime = 'video/webm; codecs="opus, vp9"';
+  var mediaSource = this;
+  var sourceBuffer = mediaSource.addSourceBuffer(mime);
+  var video = 'droid.webm';
+  fetch(video)
+    .then(function(response) {
+      <strong>return response.arrayBuffer();</strong>
+    })
+    <strong>.then(function(arrayBuffer) {
+      sourceBuffer.appendBuffer(arrayBuffer);
+    });</strong>
+}</pre>
+
+#### Call endOfStream() 
+
+At the end of processing the media file, `MediaSource.readyState` should be set
+to `ended` and a `sourceended` event should fire. Neither of these happen
+automatically, but we can trigger them by calling `MediaSource.endOfStream()`.
+Do this inside the `SourceBuffer.onupdateend`.
+
+<pre class="prettyprint">
+function sourceOpen(e) {
+  var mime = 'video/webm; codecs="opus, vp9"';
+  var mediaSource = this;
+  var sourceBuffer = mediaSource.addSourceBuffer(mime);
+  var video = 'droid.webm';
+  fetch(video)
+    .then(function(response) {
+      return response.arrayBuffer();
+    })
+    .then(function(arrayBuffer) {
+      <strong>sourceBuffer.onupdateend = function(e) {
+        mediaSource.endOfStream();
+      }</strong>
+      sourceBuffer.appendBuffer(arrayBuffer);
+    });
+}</pre>
+
+
+## The final version
+
+Here's the complete code example. I hope you have learned something about Media
```

Comment:
Tighten this up, there isn't a comment block and we don't want to encourage people to email you. Just suggest that they go to StackOverflow for questions & answers.